lib/src/optimism/mod.rs

```diff
@@ -0,0 +1,514 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::{cell::RefCell, iter::once};
+
+use alloy_sol_types::{sol, SolInterface};
+use anyhow::{bail, Context, Result};
+use serde::{Deserialize, Serialize};
+use std::collections::{HashMap, VecDeque};
+use zeth_primitives::{
+    address,
+    block::Header,
+    keccak::keccak,
+    transactions::{
+        ethereum::{EthereumTxEssence, TransactionKind},
+        optimism::{OptimismTxEssence, TxEssenceOptimismDeposited},
+        Transaction, TxEssence,
+    },
+    trie::MptNode,
+    uint, Address, BlockHash, FixedBytes, RlpBytes, B256, U256,
+};
+
+#[cfg(not(target_os = "zkvm"))]
+use log::{error, info};
+
+use crate::optimism::{
+    batcher_transactions::BatcherTransactions,
+    batches::Batches,
+    channels::Channels,
+    config::ChainConfig,
+    derivation::{BlockInfo, Epoch, State, CHAIN_SPEC},
+    epoch::BlockInput,
+};
+
+pub mod batcher_transactions;
+pub mod batches;
+pub mod channels;
+pub mod config;
+pub mod deposits;
+pub mod derivation;
+pub mod epoch;
+pub mod system_config;
+
+sol! {
+    #[derive(Debug)]
+    interface OpSystemInfo {
+        function setL1BlockValues(
+            uint64 number,
+            uint64 timestamp,
+            uint256 basefee,
+            bytes32 hash,
+            uint64 sequence_number,
+            bytes32 batcher_hash,
+            uint256 l1_fee_overhead,
+            uint256 l1_fee_scalar
+        );
+    }
+}
+
+pub trait BatcherDb {
+    fn get_full_op_block(&mut self, block_no: u64) -> Result<BlockInput<OptimismTxEssence>>;
```

Comment:
After we fix [this issue](https://github.com/risc0/zeth/pull/51#discussion_r1379689102), we can change this function so that it returns `Header` instead of `BlockInput<OptimismTxEssence>`

Reply:
Resolved by adding a new function to the trait (we still need the full block w/ transactions for the initial safe head)